core/src/util/shaderProgram.cpp

```diff
@@ -278,6 +279,24 @@ void ShaderProgram::invalidateAllPrograms() {
     
 }
 
+void ShaderProgram::setUniform(const std::string& _name, Texture& _texture) {
+    use();
+    GLint location = getUniformLocation(_name);
+    GLuint unit;
+    
+    if (m_textureSlots.find(_name) == m_textureSlots.end()) {
+        unit = m_freeTextureUnit;
+        m_textureSlots[_name] = { _texture.getGlHandle(), m_freeTextureUnit++ };
+    } else {
+        unit = m_textureSlots[_name].second;
+    }
+    
+    _texture.update(unit);
+    _texture.bind(unit);
+    
+    glUniform1i(location, unit);
+}
+
```

Comment:
I have mixed feelings about this functionality. On one hand, it's a cool convenience to essentially never worry about which texture slot you're binding to when you use a texture in a shader. On the other hand, this could lead to unintentional performance problems that are hidden behind this abstraction. As a slightly contrived example, consider a situation where two shaders are both using two textures. Shader 1 sets texture A as a uniform and then sets texture B as another uniform. Shader 2 sets texture B as a uniform and then sets texture A as another uniform. Now whenever we alternate between using shader 1 and shader 2, textures A and B will both have to be re-bound to alternate texture units (even though we could easily keep the same bindings if we were a little more careful). If we were explicitly binding the textures this would be easy to see. 

PS: I didn't intend to think that much about this PR but it actually brings up some really interesting issues!